### DISCUSSION TITLE
Experimental: patch conversions

### DISCUSSION THREAD

**Discussion by BobanSpasic** [probonopd/MiniDexed#560] [Show and tell]
Edit 02.12.2023:
The conversion software is now located here: https://github.com/BobanSpasic/MDX_PerfConv
The ready-to-use files (successful conversion) are here: https://github.com/BobanSpasic/MDX_Vault

=================================================================
Old post:

**Downloads removed because the conversion wasn't good. Please, read the thread further for the details**

It is a batch conversion, I didn't really check how it sounds.
I am working on a converter. Please report bugs (Detune, NoteShift, SplitPoints).
Do not expect wonders - these are conversions from DX7II. It means, we miss much of the special effects found on DX7II (Stereo PAN envelope, extra modulation assigns etc.)
In Yamaha Freeware package, the performances 193 to 200 have 8 voices each assigned to MIDI channels 1 to 8.

DX5.zip, DX7IIFD.zip and DX7s.zip are factory presets. You will see that DX7s has no dual or split performances, just singles - it isn't an error.

**Comment by BobanSpasic** [comment]
OK, I've found two major errors 
 - split points are overlapping 2 tones
 - that slide on almost every performance... grrr.. can someone tell me which parameter does that?

**Reply by probonopd** [reply]
For example, if you want middle C as the split point:

```
# TG1
NoteLimitLow1=0
NoteLimitHigh1=59

# TG2
NoteLimitLow2=60
NoteLimitHigh2=127
```

**Reply by probonopd** [reply]
We must document each parameter in https://github.com/probonopd/MiniDexed/wiki/Files#performanceini. Volunteers?

**Comment by BobanSpasic** [comment]
>     * that slide on almost every performance... grrr.. can someone tell me which parameter does that?
Pitch EG Level/Rate.
Dexed has the same problem, it is also sliding. I've compared to FM7, and there is OK.

**Reply by probonopd** [reply]
Isn't that a _voice_ thing, so unrelated to performances? (Not 100% sure)

**Reply by BobanSpasic** [reply]
It is a voice thing, but it seems there are differences:  
http://www.muzines.co.uk/articles/decisions-decisions/2058  
(scroll over the middle of the page)

This is the python code from DXConvert that does the translation (comments at the end of the lines are from me):  
```
 # PEG range DX7II AMEM to DX7 VMEM conversion:
    if dx72==False:
        for p in range(len(dx7data)//128):  #split VMEM to single voices
            for i in range(106, 110):		#Pitch EG Level
                pegr=dx72data[35*p+4]&3		# PEGR from AMEM, 0-3
                peg=dx7data[128*p+i]
                # dx7data[128*p+i]=int((peg-50)*(50, 32, 18, 10)[pegr]/50.0) + 50
                dx7data[128*p+i]=int((peg-50)*(50, 32, 16, 8)[pegr]/50.0) + 50 #need help to understand the formula

    # DX7 "Rocksplit" correction
    rocksplit = (100, 102, 103, 104, 106, 107, 108, 109, 110, 115,
                  116, 118, 120, 121, 122, 123, 125, 126, 127)
    if dx72 or TX7:
        for p in range(len(dx7data)//128):
            for op in range(6):
                if dx7data[128*p + 14 + 17*op] in rocksplit: #TL
                    if dx7data[128*p + 11 + 17*op] >> 2 == 0: #RC
                        dx7data[128*p + 10 + 17*op] = 99 #RD
                    if dx7data[128*p +11 + 17*op] & 3 == 0: #LC
                        dx7data[128*p + 9 + 17*op] = 99
                    if dx72:
                        dx7data[128*p + 14 + 17*op] = 99 + 16 * (99 - dx7data[128*p + 14 + 17*op]) // 28
```

**Reply by probonopd** [reply]
Wow. This whole Yamaha data thing is way more complex than what meets the eye at the surface!

**Reply by probonopd** [reply]
On an unrelated note, @BobanSpasic is there any way to get you away from using Pascal, a language that very few developers know these days? At least for the "backend"/"business logic" stuff like these conversions. I think if we could use C++, then we might even be able to put the conversions/parsing into MiniDexed/Dexed/... one day.

**Reply by BobanSpasic** [reply]
> On an unrelated note, @BobanSpasic is there any way to get you away from using Pascal, a language that very few developers know these days? At least for the "backend"/"business logic" stuff like these conversions. I think if we could use C++, then we might even be able to put the conversions/parsing into MiniDexed/Dexed/... one day.

You can't teach an old horse to plow (Balkan version of "you can't teach an old dog new tricks")
You won't believe how much software is developed in Free Pascal or Delphi, but nobody want to speak about it, just like hiding a curse or something like that. I am programming PLCs for living. The standard programming language for PLCs goes under the name ST (Structured Text) - it is just a specialized variant of Pascal.

I'll get the conversion from DX7II to DX7 finished tomorrow, and I'll check back here with the results.

**Reply by Banana71** [reply]
> It is a voice thing, but it seems there are differences:
> http://www.muzines.co.uk/articles/decisions-decisions/2058
> (scroll over the middle of the page)

Oops, I wasn't aware of the problems with the delay effect until now.

I have also always overlooked the page in the DX7II operating instructions. Sad that Dexed is based on the DX7 MK I.
![Screenshot 2023-11-17 100744](https://github.com/probonopd/MiniDexed/assets/104296205/5094d183-ae75-41b9-8d3f-99a64c573f3e)

**Comment by Banana71** [comment]
> OK, I've found two major errors
> 
> * split points are overlapping 2 tones
> * that slide on almost every performance... grrr.. can someone tell me which parameter does that?

I was surprised that aftertouch didn't work and then I found this:
the converter:

```
AfterTouchRange1=65
AfterTouchTarget1=4
```
correct would be:
```
AftertouchRange1=65
AftertouchTarget1=4
```

**Reply by BobanSpasic** [reply]
Thanks Peter. You mean, it is case-sensitive?

**Reply by Banana71** [reply]
> You mean, it is case-sensitive?

Yes, the miniDexed seems to make a difference. Only when I changed and saved the parameters by hand could the aftertouch effect be heard

**Comment by BobanSpasic** [comment]
Take 2:
- Split point correction
- Aftertouch correction
- Pitch EG translation for DX7II/TX802 voices
- Amplitude Modulation Sensitivity translation for DX7II/TX802 voices
- Detune correction for DX7II/TX802 voices

Test-bed samples:
[DX7II_Yamaha_Freeware.zip](https://github.com/probonopd/MiniDexed/files/13390146/DX7II_Yamaha_Freeware.zip)

It is not perfect, but it is playable

btw. in the ZIP file are a couple of performances using Breath Controller - what should we do about it?

**Reply by Banana71** [reply]
Thanks @BobanSpasic Boban, the performances are very well converted.

>a couple of performances using Breath Controller - what should we do about it?

Maybe @diyelectromusic Kevin can design a breath controller and integrate it directly into the miniDexed?
https://www.youtube.com/watch?v=es92GqYBge0

**Reply by BobanSpasic** [reply]
@Banana71 - Thanks for reporting.

The repository for test files is here: https://github.com/BobanSpasic/MDX_PerfConv/tree/main/Test_files
Currently, the Coffeeshoped TX802 files does not convert at all. I need to take a look at the untDispatcher code.
There is also missing B1 and B2 banks from TX802. I can't find the original files on the net. I can only find the VMEM dumps (4104 bytes), but these are totally wrong.
For DX5 factory files, there is no other source other than 12-op's conversion to DX7II.
ChiNoMaki and Ajay collections are also in wrong formats, without AMEM data.

If anyone has some more interesting "other-than-DX7" files, and want to share - please do so :)

**Comment by BobanSpasic** [comment]
I think that I've finished MDX_PerfConv, except if some bugs occur.
I ordered some CDs with voices for all the DX/TX 6-OP models, inclusive Factory presets. It should arrive this or the next week. Let's see how the MDX_PerfConv makes out with it.

**Reply by BobanSpasic** [reply]
Well, it is not finished yet - I've just discovered I could also convert PLG150-DX w2b banks.
I've just found out, that a PLG150-DX dump from Yamaha Motif ES contains two DX7II SysEx-es in the w2b file.
http://www.motifator.com/index.php/forum/viewthread/453750/

```
MDX_PerfConv.exe -c -a NOSEBLEE.W2B -v
Output directory: c:\Users\Bobby\Documents\Lazarus\MiniDexed_Projects\MDX_PerfConv\Test_files\PLG150DX\NOSEBLEE.W2B
It is a multiple DX7II bank with supplement
Using ConvertMultiDX7IItoMDX with one stream
VMEM loaded from 3727
Nosebleed
KlingKlang
Swimming
Ethnectric
Percussive
Skirmish
Still DX
Stomper
Wreath
Subculture
Ni-Chi-Bu
KAGERO
Warp
Energie
Clokker
GB Trance
Hypnotic
RaggaGabba
Nose Dive
Sink
Logi Drum
Bombay
Oaz
BrasilDrum
Asiara
Acid Pulse
Elec Pulse
Dredzs
Elec Rezz
Night Life
Orbo Fuzz
Big Up
AMEM loaded from 7831
VMEM loaded from 8966
Drenched
TunaMooger
Digi Touch
Floaty
Cyanine
Pingpong
Mini Jim
Distcore
Hardcore
HouseParty
Ibiza
Sand
ChimeNoise
Deep Beat
Full House
E-Salsa
Scream
FJA
Minimal At
Voodoo Key
Oskylaytor
Stupid
Solo Bass
Saw FB
Badass
Quid
Reeper
Scratches
Fuzz Line
House Dog
FreeGroove
Life Line
AMEM loaded from 13070
```

**Reply by BobanSpasic** [reply]
> I think that I've finished MDX_PerfConv, except if some bugs occur. I ordered some CDs with voices for all the DX/TX 6-OP models, inclusive Factory presets. It should arrive this or the next week. Let's see how the MDX_PerfConv makes out with it.


 CD arrived, but... it is full of files that we already have from the internet, and with files in Sound Diver 3 format (ancient software, and not free). I need to convert them to SysEx first...

**Reply by probonopd** [reply]
https://github.com/BobanSpasic/MDX_Vault/tree/main/DX7II/ROMs - do these correspond to the "VRC Voice ROMs" at https://yamahablackboxes.com/collection/yamaha-dx7ii-fd-synthesizer/patches/? If so, do they really contain performances or just collections of voices?

**Reply by BobanSpasic** [reply]
Yes, these are the VRCs, and yes - I have the complete cartridges inclusive performances.
See a demo from 12-op here: https://www.youtube.com/watch?v=DKTSdxaUPIc
On 12-op's channel, you can find demos from other VRCs too.

btw. about the demos - DX7II has advanced pan control, MiniDexed don't. Do not expect the same sound as in the demo.

**Reply by probonopd** [reply]
> DX7II has advanced pan control, MiniDexed don't

Can you point me to a description of what is missing? Maybe we can add it to MiniDexed?

**Reply by BobanSpasic** [reply]
First, you have two outputs (I and II) - MiniDexed too (stereo, L and R), and that's the end of the similarities.
After that, - DX7II has pan modes and pan envelopes. 
[DX7IIE.PDF](https://github.com/probonopd/MiniDexed/files/13538817/DX7IIE.PDF)
Pages 36 and 37
See, you can write envelopes, how the voice moves between the outputs, or how the levels of the two voices can change with time.
Also, you can assign this Pan controls to ModWheel, or to Aftertouch etc.

**Reply by Banana71** [reply]
>Can you point me to a description of what is missing? Maybe we can add it to MiniDexed?

see here #407

**Reply by probonopd** [reply]
Wow. The learning never ends in the magical world that is 80s FM. I opened ticket #576 to discuss this topic further.

**Reply by probonopd** [reply]
Comparing the names of the performances in the converted VRCs, it is not clear to me how they map to the original as seen in the videos @BobanSpasic. Did you convert the 1001,....1005? Maybe you can help me fill in the blanks on the [Performances wiki page](https://github.com/probonopd/MiniDexed/wiki/Performances)?

**Reply by BobanSpasic** [reply]
I don't know the VRC numbers of the cartridges.
At the moment, I am trying to fix the bugs in MDX_PerfConv and publish the converted files. I can't promise I'll have time for the Wiki today.
btw. the guy's nickname is 12-op, not 7-op (DX7II has 2 voices x 6 operators, thus 12-op)

**Reply by probonopd** [reply]
We are not in a hurry, take your time. I hope all of this is as much fun for you to make these as it is for me to try them.

**Comment by BobanSpasic** [comment]
[TX802 ROM Decay.zip](https://github.com/probonopd/MiniDexed/files/13504113/TX802.ROM.Decay.zip)
Is the Detune translation OK, or it is too much/high?

**Reply by probonopd** [reply]
Thanks for your hard work and dedication @BobanSpasic.
I guess only someone with access to a real TX802 can say.

**Comment by BobanSpasic** [comment]
Good news - I've reconstructed the whole ROM of the TX802 (all 128 voices). You can't find it on the internet. It was a hell of the job, but it is now completed.
Soon, we will have the correct translations of the TX802 performances.

**Reply by diyelectromusic** [reply]
Oh, jolly well done :)

**Reply by BobanSpasic** [reply]
Ladies and gentlemen, ta-daaaaaaa:
https://github.com/BobanSpasic/MDX_Vault

Added: TX802 Factory performances, DX7II Cartridges
TX802 Cartridges are the same as DX7II, so I didn't add them. I didn't find TX802 cartridges that are going beyond DX7II capabilities (two TGs).
Only difference for converted cartridges is, that TX802 performances are using TG1 and TG5, where DX7II performances are using TG1 and TG2.

Differences between TX802 Factory and the files from fmori (I do not know the real name, so I am using the nickname from the mailing list):
A bit of theory before I can explain the difference - one TX802 Performance has 8 TGs (voices), where each voice in the performance is defined (selected from the voice banks), but it does not mean it will also be used in the performance. Namely, TX802 has "link" function for TGs, where more TGs are borrowing polyphony to one TG.
The voices from the linked TGs are not used, because these TGs are playing the voice from the "main" TG.
E.g (TX802 display).  
` A-02 <-- <-- <-- B-08 <-- <-- <--`  
 means that TG1 is playing voice A-02, the next 3 TGs are borrowing polyphony to TG1. TG5 is playing the voice B-08 and the next 3 TGs are borrowing polyphony to TG5. Even if the 6 TGs are just borrowing polyphony, the voices for these TGs are also selected in the performance.
The selected voices are included in the performance.ini files, but the MIDI channel is set to 0 to mute them
Back to the difference between the fmori and other factory performance - the muted voices are different. Original has INIT VOICE for the muted linked voices, and in fmori files are voices from his Internal Voices.

**Comment by BobanSpasic** [comment]
First post updated with the links

**Comment by probonopd** [comment]
With so many performances to choose from thanks to the great work of @BobanSpasic, we need a simpler way to deal with more performance files.

I have outlined what I have in mind in https://github.com/probonopd/MiniDexed/pull/575 (but currently it is pseudo code). Maybe @diyelectromusic can have a look at it?

**Comment by BobanSpasic** [comment]
DX5 factory performance file reconstructed from two corrupted files.
The conversion has bugs. Just the first 32 performances are converted, the rest is INIT.
I'll check the code tomorrow, now I need a sleep...

**Reply by BobanSpasic** [reply]
Done

**Comment by BobanSpasic** [comment]
Sorry, but I needed to update the DX5 directory again. Now are the NoteShift parameters implemented like it should.
I've added Pan to every performance, but this is superficial, because there is no information to translate from the original SysEx.
For dual voices, I've set the Pan to 32 and 96, and for splits to 0 and 127 (hard left and hard right).

There are a lot of undocumented parameters in DX5 performance SysEx, but unfortunately I can't find any editor/librarian that supports them, so that I can do some investigation. Other solution were a real DX5, but this is even less possible.
